ycmd/handlers.py

```diff
@@ -89,17 +90,39 @@ def RunCompleterCommand():
 def GetCompletions():
   _logger.info( 'Received completion request' )
   request_data = RequestWrap( request.json )
-  do_filetype_completion = _server_state.ShouldUseFiletypeCompleter(
-    request_data )
+  ( do_filetype_completion, forced_filetype_completion ) = (
+                    _server_state.ShouldUseFiletypeCompleter(request_data ) )
   _logger.debug( 'Using filetype completion: %s', do_filetype_completion )
-  filetypes = request_data[ 'filetypes' ]
-  completer = ( _server_state.GetFiletypeCompleter( filetypes ) if
-                do_filetype_completion else
-                _server_state.GetGeneralCompleter() )
 
-  return _JsonResponse( BuildCompletionResponse(
-      completer.ComputeCandidates( request_data ),
-      request_data.CompletionStartColumn() ) )
+  general_completer = _server_state.GetGeneralCompleter()
+  kw = {}
+
+  if do_filetype_completion:
+    try:
+      completions = ( _server_state.GetFiletypeCompleter(
+                                  request_data[ 'filetypes' ] )
+                                 .ComputeCandidates( request_data ) )
+
+      if not completions and not forced_filetype_completion:
+        completions = general_completer.ComputeCandidates( request_data )
+    except Exception as e:
+      if not forced_filetype_completion:
+        completions = general_completer.ComputeCandidates( request_data )
```

Comment:
This repeats the code from above. I think you can restructure this to avoid the repetition.

Maybe something like:
```
completions = None
if do_filetype_completion:
  try:
    completions = <get semantic>
  except:
    if forced:
      raise
    else:
      <store error>
if not completions and not forced:
  completions = <get ident>
return completions + error data if any
```